lib/puppet/provider/package/pip.rb

```diff
@@ -166,38 +202,69 @@ def latest_with_old_pip
       execpipe command_and_options do |process|
         process.collect do |line|
           # PIP OUTPUT: Using version 0.10.1 (newest of versions: 1.2.3, 4.5.6)
-          if line =~ /Using version (.+?) \(newest of versions/
-            return $1
+          if line =~ /Using version .+? \(newest of versions: (.+?)\)/
+            textAfterLastMatch = $1.chomp(")\n")
+            versionList = textAfterLastMatch.split(', ').sort do |x,y|
+              Puppet::Util::Package.versioncmp(x, y)
+            end
+            return versionList
           end
         end
-        return nil
+        return []
       end
     end
   end
 
+  # Finds the most suitable version available in a given range
+  def best_version(should_range)
+    included_available_versions = []
+    available_versions.each do |version|
+      version = PIP_VERSION.parse(version)
+      included_available_versions.push(version) if should_range.include?(version)
+    end
+
+    sorted_result = included_available_versions.sort { |x,y| x <=> y }
+    return sorted_result.to_a.last if sorted_result
+
+    Puppet.debug("No available version for package #{@resource[:name]} is included in range #{should_range}")
+    should_range
+  end
+
   # Install a package.  The ensure parameter may specify installed,
   # latest, a version number, or, in conjunction with the source
   # parameter, an SCM revision.  In that case, the source parameter
   # gives the fully-qualified URL to the repository.
-
   def install
     command = resource_or_provider_command
     self.class.validate_command(command)
 
+    should = @resource[:ensure]
     command_options = %w{install -q}
     command_options +=  install_options if @resource[:install_options]
     if @resource[:source]
-      if String === @resource[:ensure]
-        command_options << "#{@resource[:source]}@#{@resource[:ensure]}#egg=#{@resource[:name]}"
+      if String === should
+        command_options << "#{@resource[:source]}@#{should}#egg=#{@resource[:name]}"
       else
         command_options << "#{@resource[:source]}#egg=#{@resource[:name]}"
       end
     else
-      case @resource[:ensure]
-      when String
-        command_options << "#{@resource[:name]}==#{@resource[:ensure]}"
+      case should
       when :latest
         command_options << "--upgrade" << @resource[:name]
+      when String
+        begin
+          should_range = PIP_VERSION_RANGE.parse(should, PIP_VERSION)
+          should = best_version(should_range)
+
+          if should == should_range
```

Comment:
The scope here was for when no version that satisfies the given range is found. We agreed, for now at least, that we will let the provider handle this case and give the error output. 

We need to concatenate `pip install package` with either `==A` or the introduced operators such as `>A <B`. Without this if, we would end up with something like `==>A <B` which will throw a syntax error rather than telling the user that no package version in the given range was found.  

I opted to replace  `should == should_range` with `should.is_a?(PIP_VERSION_RANGE)`. I think this is more readable.